### ISSUE TITLE
A Generalized Weisfeiler-Lehman Graph Kernel

### ISSUE THREAD

**Issue by hsanchez** [hsanchez/paper-memory#5]
## TL;DR

A generalization of the Weisfeiler-Lehman graph kernels, which takes into account the similarity between unfolding trees rather than their equality during kernel generation.

### Paper Link

- [arxiv.org/abs/2101.08104](https://arxiv.org/abs/2101.08104)
- [github.com/mlai-bonn/GenWL](https://github.com/mlai-bonn/GenWL)

### Author/Institution

- Till Hendrik Schulz (University of Bonn, Germany)
- Tamas Horvath (Fraunhofer IAIS, Sankt Augustin, Germany)
-  Pascal Welke (University of Bonn, Germany)
-  Stefan Wrobel (Fraunhofer IAIS, Sankt Augustin, Germany)

## Overview

Despite their noticeable speed, WeisfeilerLehman graph kernels are conceptually limited to comparing labels. While this comparison is extremely well-suited for graph isomorphism tests, it is arguably too restrictive for defining similarities, in particular, graph kernels.  E.g., let's consider the unfolding trees depicted below, while _T1_ visibly resembles _T2_ much more than it resembles _T3_, the Weisfeiler-Lehman graph kernel treats them all as unequal and is thus unable to _quantify_ the apparent differences among pairwise similarities between the unfolding trees.

```mermaid
graph TD

subgraph T3
    1(( ))-->2(( ))
    1-->3(( ))
    1-->4(( ))
    2-->5(( ))
    2-->6(( ))
    2-->7(( ))
    3-->8(( ))
    4-->9(( ))
    4-->10(( ))
end
subgraph T2
    E(( ))-->F(( ))
    F-->G(( ))
    F-->H(( ))
    F-->I(( ))
end
subgraph T1
    A(( ))-->B(( ))
    B-->C(( ))
    B-->D(( ))
end
```

Motivated by these considerations, the authors relax the above strictness by proposing a method that compares Weisfeiler-Lehman labels, or equivalently unfolding trees, with regard to a much finer similarity measure than the binary-valued one.

Experimental results show that the proposed approach considerably outperforms state-of-the-art kernels (including the ordinary Weisfeiler-Lehman subtree kernel) on datasets containing dense and structurally diverse graphs.

## Contributions and Distinctions from Previous Works

Prior works' vertex matching abilities are limited by the rigid Weisfeiler-Lehman label refinement method. While the idea of comparing vertices by a finer similarity measure than the equivalence of their neighborhoods has been considered already, (1) the DAGs considered by this approach describe structurally different representations of neighborhoods, (2) the similarity between graphs is based on the concept of tree edit distances rather than the one concerning tree kernels.

## Methods

The proposed approach employs a similarity between Weisfeiler-Lehman labels based on the concept of tree edit distances between their respective unfolding trees. **The key concept of the relaxed Weisfeiler-Lehman subtree kernel is to identify groups of similar Weisfeiler-Lehman labels by clustering**. The elements within a cluster are then treated as identical labels. That is, we generalize the ordinary Weisfeiler-Lehman kernel by regarding two unfolding trees equivalent if they belong to the same cluster, i.e., if they have a small distance to each other. In this way, the ordinary Weisfeiler-Lehman kernel is the special case where labels are considered equivalent only if they have a distance of zero. With regards to the partitioning of Wesfeiler-Lehman labels, the authors use the Wasserstein k-means clustering algorithm introduced by Irpino's paper on _Dynamic clustering of histogram data based on adaptive squared Wasserstein distances_.

## Results

No visible performance improvements on datasets containing mainly sparse and noise-free graphs such as molecular data. However, results show that this approach outperformed all other considered kernels on datasets containing fuzzy and structurally diverse graphs.

The original Weisfeiler-Lehman kernel is only suitable when there are only a few different unfolding trees in the graphs of the dataset such that these graphs share sufficiently many labels in order to compute meaningful similarities. The proposed approach makes up for this drawback. It is able to identify similar vertex neighborhoods, leading to major increases in predictive performance on datasets containing noisy and structurally diverse graphs.
